src/drivers/imx/sdma.c

```diff
@@ -0,0 +1,848 @@
+	int ret = sdma_register_init(dma);
+	if (ret < 0)
+		return ret;
+	sdma_init_c0(dma);
+	trace_sdma_error("sdma_boot done");
+	return 0;
+}
+
+static int sdma_upload_context(struct dma_chan_data *chan) {
+	struct sdma_chan *pdata = dma_chan_get_data(chan);
+
+	/* Ensure context is ready for upload */
+	dcache_writeback_region(pdata->ctx, sizeof(*pdata->ctx));
+	trace_sdma_error("sdma_upload_context for ch %d", chan->index);
+
+	/* Last parameters are unneeded for this command and are ignored;
+	 * set to 0.
+	 */
+	return sdma_run_c0(chan->dma, SDMA_CMD_C0_SET_DM, (uint32_t)pdata->ctx,
+			   SDMA_SRAM_CONTEXTS_BASE + chan->index * sizeof(*pdata->ctx) / 4,
+			   sizeof(*pdata->ctx) / 4);
+}
+
+__attribute__((unused)) // TODO use
+static int sdma_download_context(struct dma_chan_data *chan) {
+	struct sdma_chan *pdata = dma_chan_get_data(chan);
+
+	/* Last parameters are unneeded for this command and are ignored;
+	 * set to 0.
+	 */
+	trace_sdma_error("sdma_download_context channel %d", chan->index);
+	int ret = sdma_run_c0(chan->dma, SDMA_CMD_C0_GET_DM, (uint32_t)pdata->ctx,
+			      SDMA_SRAM_CONTEXTS_BASE + chan->index * sizeof(*pdata->ctx) / 4,
+			      sizeof(*pdata->ctx) / 4);
+	dcache_invalidate_region(pdata->ctx, sizeof(*pdata->ctx));
+
+	return ret;
+}
+
+static int sdma_upload_contexts_all(struct dma *dma) {
+	struct sdma_pdata *pdata = dma_get_drvdata(dma);
+
+	trace_sdma_error("sdma_upload_contexts_all");
+	dcache_writeback_region(pdata->contexts, sizeof(*pdata->contexts));
+
+	 /* Division by 4 in size calculation is because count is in words and
+	  * not in bytes
+	  */
+	return sdma_run_c0(dma, SDMA_CMD_C0_SET_DM, (uint32_t)pdata->contexts,
+			   SDMA_SRAM_CONTEXTS_BASE,
+			   dma->plat_data.channels * sizeof(*pdata->contexts) / 4);
+}
+
+static int sdma_download_contexts_all(struct dma *dma) {
+	struct sdma_pdata *pdata = dma_get_drvdata(dma);
+
+	trace_sdma_error("sdma_download_contexts_all");
+	int ret = sdma_run_c0(dma, SDMA_CMD_C0_GET_DM, (uint32_t)pdata->contexts, SDMA_SRAM_CONTEXTS_BASE, dma->plat_data.channels * sizeof(*pdata->contexts) / 4); /* Division by 4 because count is in words and not in bytes */
+
+	dcache_invalidate_region(pdata->contexts, sizeof(*pdata->contexts));
+
+	return ret;
+}
+
+// TODO
+static void sdma_teardown_contexts(struct dma *dma) {
+	trace_sdma_error("SDMA: Contexts teardown (free path), TODO");
+}
+
+// TODO
+static void sdma_shutdown(struct dma *dma) {
+	trace_sdma_error("SDMA: Shutdown, TODO");
+}
+
+static void irqhdnlr(void *data) {
+	trace_sdma_error("SDMA interrupt works");
+}
+
+/* Below SOF related functions will be placed */
+
+static int sdma_probe(struct dma *dma)
+{
+	int channel;
+	int ret;
+	struct sdma_pdata *pdata;
+
+	interrupt_register(interrupt_get_irq(dma_irq(dma), dma_irq_name(dma)), irqhdnlr, dma);
+
+	if (dma->chan) {
+		trace_sdma_error("SDMA: Repeated probe");
+		return -EEXIST;
+	}
+
+	trace_sdma("SDMA: probe");
+	trace_sdma_error("sdma_probe");
+
```

Comment:
Just keep trace_sdma here.

Reply:
The trace_sdma_error calls will all go (some may be converted to trace or trace_verbose if they are important enough)

Reply:
thanks.